File: scripts/validate-spec.py
#!/usr/bin/env python3
"""
Validate spec files for required fields and format.

Usage: python validate-spec.py <path-to-spec.md>
       python validate-spec.py --all --specs-dir specs/
"""

import argparse
import re
import sys
from pathlib import Path


REQUIRED_FIELDS = ['title', 'status', 'domain', 'issue', 'created', 'updated']
VALID_STATUSES = ['active', 'deprecated', 'superseded', 'archived']


def parse_frontmatter(content: str) -> dict | None:
    """Extract YAML frontmatter from markdown."""
    match = re.match(r'^---\s*\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return None

    frontmatter = {}
    for line in match.group(1).split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            frontmatter[key.strip()] = value.strip()
    return frontmatter


def validate_spec(path: Path) -> list[str]:
    """Validate a spec file. Returns list of errors."""
    errors = []

    if not path.exists():
        return [f"File not found: {path}"]

    content = path.read_text()

    fm = parse_frontmatter(content)
    if fm is None:
        return [f"Missing frontmatter in {path}"]

    for field in REQUIRED_FIELDS:
        if field not in fm or not fm[field]:
            errors.append(f"Missing required field '{field}' in {path}")

    if 'status' in fm and fm['status'] not in VALID_STATUSES:
        errors.append(f"Invalid status '{fm['status']}' in {path}. Must be one of: {VALID_STATUSES}")

    if 'issue' in fm:
        issue = fm['issue']
        if issue in ['PROJ-XXX', '[PROJ-XXX]', 'TODO', '', '{{ISSUE}}']:
            errors.append(f"Issue field is placeholder in {path}. Must reference actual issue.")

    return errors


def main():
    parser = argparse.ArgumentParser(description='Validate spec files')
    parser.add_argument('path', nargs='?', help='Path to spec file')
    parser.add_argument('--all', action='store_true', help='Validate all specs')
    parser.add_argument('--specs-dir', default='specs/', help='Specs directory')
    args = parser.parse_args()

    if args.all:
        specs_dir = Path(args.specs_dir)
        specs = [p for p in specs_dir.rglob('*.md')
                 if p.name not in ('INDEX.md', 'SNAPSHOT.md', 'glossary.md')]

        all_errors = []
        for spec in specs:
            errors = validate_spec(spec)
            all_errors.extend(errors)

        if all_errors:
            print("Validation errors:")
            for error in all_errors:
                print(f"  - {error}")
            return 1

        print(f"✓ All {len(specs)} specs are valid")
        return 0

    elif args.path:
        errors = validate_spec(Path(args.path))
        if errors:
            print("Validation errors:")
            for error in errors:
                print(f"  - {error}")
            return 1
        print(f"✓ {args.path} is valid")
        return 0

    else:
        parser.print_help()
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: plugins/sdd/scripts/validate-spec.ts
#!/usr/bin/env npx ts-node --esm
/**
 * Validate spec files for required fields and format.
 *
 * Usage: npx ts-node --esm validate-spec.ts <path-to-spec.md>
 *        npx ts-node --esm validate-spec.ts --all --specs-dir specs/
 */

import * as fs from 'node:fs/promises';
import { parseFrontmatter } from './lib/frontmatter.js';
import { findSpecFiles, directoryExists } from './lib/spec-utils.js';

const REQUIRED_FIELDS = ['title', 'status', 'domain', 'issue', 'created', 'updated'] as const;
const VALID_STATUSES = ['active', 'deprecated', 'superseded', 'archived'] as const;
const PLACEHOLDER_ISSUES = ['PROJ-XXX', '[PROJ-XXX]', 'TODO', '', '{{ISSUE}}'] as const;

interface ValidationError {
  readonly file: string;
  readonly message: string;
}

/**
 * Validate a spec file. Returns list of errors.
 */
const validateSpec = async (specPath: string): Promise<readonly ValidationError[]> => {
  const errors: ValidationError[] = [];

  try {
    await fs.access(specPath);
  } catch {
    return [{ file: specPath, message: 'File not found' }];
  }

  const content = await fs.readFile(specPath, 'utf-8');
  const fm = parseFrontmatter(content);

  if (!fm) {
    return [{ file: specPath, message: 'Missing frontmatter' }];
  }

  for (const field of REQUIRED_FIELDS) {
    if (!fm[field]) {
      errors.push({ file: specPath, message: `Missing required field '${field}'` });
    }
  }

  const status = fm['status'];
  if (status && !(VALID_STATUSES as readonly string[]).includes(status)) {
    errors.push({
      file: specPath,
      message: `Invalid status '${status}'. Must be one of: ${VALID_STATUSES.join(', ')}`,
    });
  }

  const issue = fm['issue'];
  if (issue && (PLACEHOLDER_ISSUES as readonly string[]).includes(issue)) {
    errors.push({
      file: specPath,
      message: 'Issue field is placeholder. Must reference actual issue.',
    });
  }

  return errors;
};

/**
 * Parse command line arguments.
 */
const parseArgs = (args: readonly string[]): { all: boolean; specsDir: string; path?: string } => {
  let all = false;
  let specsDir = 'specs/';
  let specPath: string | undefined;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--all') {
      all = true;
    } else if (arg === '--specs-dir') {
      specsDir = args[i + 1] ?? 'specs/';
      i++;
    } else if (!arg?.startsWith('-')) {
      specPath = arg;
    }
  }

  return { all, specsDir, path: specPath };
};

const main = async (): Promise<number> => {
  const args = parseArgs(process.argv.slice(2));

  if (args.all) {
    if (!(await directoryExists(args.specsDir))) {
      console.error(`Error: Specs directory not found: ${args.specsDir}`);
      return 1;
    }

    const specs = await findSpecFiles(args.specsDir);
    const allErrors: ValidationError[] = [];

    for (const spec of specs) {
      const errors = await validateSpec(spec.path);
      allErrors.push(...errors);
    }

    if (allErrors.length > 0) {
      console.log('Validation errors:');
      for (const error of allErrors) {
        console.log(`  - ${error.file}: ${error.message}`);
      }
      return 1;
    }

    console.log(`✓ All ${specs.length} specs are valid`);
    return 0;
  }

  if (args.path) {
    const errors = await validateSpec(args.path);
    if (errors.length > 0) {
      console.log('Validation errors:');
      for (const error of errors) {
        console.log(`  - ${error.message}`);
      }
      return 1;
    }
    console.log(`✓ ${args.path} is valid`);
    return 0;
  }

  console.log('Usage: validate-spec.ts <path-to-spec.md>');
  console.log('       validate-spec.ts --all --specs-dir specs/');
  return 1;
};

main()
  .then(process.exit)
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
